File: no-starve-mutex/python/noStarveMutex.py
import threading
import psutil

class RoomInit:
    room1 = 0
    room2 = 0
    mutex = threading.Semaphore(1)
    t1 = threading.Semaphore(1)
    t2 = threading.Semaphore(0)

def waitingRooms(roomInit, threadId):
    roomInit.mutex.acquire()
    roomInit.room1 += 1
    print "Thread#%d is in room1." % (threadId)
    roomInit.mutex.release()

    roomInit.t1.acquire()
    roomInit.room2 += 1
    roomInit.mutex.acquire()
    roomInit.room1 -= 1
    print "Thread#%d is in room2." % (threadId)

    if roomInit.room1 == 0:
        print "room1 is empty now."
        roomInit.mutex.release()
        roomInit.t2.release()
    else:
        roomInit.mutex.release()
        roomInit.t1.release()

    roomInit.t2.acquire()
    roomInit.room2 -= 1

    if roomInit.room2 == 0:
        print "room2 is empty now."
        roomInit.t1.release()
    else:
        roomInit.t2.release()

def runThreads():
    roomInit = RoomInit()

    N_THREADS = 1000

    t = [None] * N_THREADS

    for i in range (0, N_THREADS):
        t[i] = threading.Thread(target=waitingRooms, args=[roomInit, i])
        t[i].start()
    
    for i in range (0, N_THREADS):
        t[i].join()
        
        
    print "CPU: ", psutil.cpu_percent()
    

runThreads()

File: no-starve-mutex/java/NoStarveMutex.java
import java.util.concurrent.Semaphore;

class WaitRooms {
    int room1 = 0;
    int room2 = 0;

    static Semaphore mutex = new Semaphore(1);
    static Semaphore t1 = new Semaphore(1);
    static Semaphore t2 = new Semaphore(0);

    void runThoughWaitingRooms(int threadId) {
        try {
            mutex.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        } 
        room1 ++;
        System.out.println("Thread#" + threadId + " is in room1.");
        mutex.release();

        try {
            t1.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        }
        room2 ++;
        try {
            mutex.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        } 
        room1 --;

        if (room1 == 0) {
            System.out.println("Room1 is empty now.");
            mutex.release();
            t2.release();
        } else {
            mutex.release();
            t1.release();
        }

        try {
            t2.acquire();
        } catch(InterruptedException e) { 
            System.out.println("InterruptedException caught"); 
        } 
        room2 --;

        System.out.println("Thread#" + threadId + " is in room2.");

        if (room2 == 0) {
            System.out.println("Room2 is empty now.");
            t1.release();
        } else {
            t2.release();
        }
    }
}

class ClientThread implements Runnable{
    WaitRooms waitRooms;
    int threadId;
    
    ClientThread(WaitRooms waitRooms, int threadId) {
        this.waitRooms = waitRooms;
        this.threadId = threadId;

        new Thread(this, "ClientThread").start();
    }

    public void run() {
        this.waitRooms.runThoughWaitingRooms(this.threadId);
    }
}

public class NoStarveMutex {
    public static void main(String[] args) {
        WaitRooms waitRooms = new WaitRooms();

        for (int i = 0; i < 1000; i++) {
            new ClientThread(waitRooms, i);
        }
    }
}